### PR TITLE
fix(home): crossfade the portfolio balance into the top bar as the header collapses

### LINKED ISSUE #5013: Home: portfolio balance visible twice while the header collapses — needs a scroll-driven crossfade
Scrolling the vault home moves the portfolio balance into the top bar. Mid-transition **both are legible at once** — the header balance is fully drawn while the large balance is still visible below the header edge. Looks like a duplicate, not a transition. iOS + macOS, Wallet + DeFi tabs.

**Cause:** binary threshold, and the large balance isn't tied to it.
- `VaultMainScreen.swift:260-263` — `showBalanceInHeader = offset < contentInset` (Bool, not progress).
- `HomeMainHeaderView.swift:45-48,53-58` — spring-animates its own swap on its own timeline.
- `VaultMainScreen.swift:143` — `VaultMainBalanceView` has no opacity/scale bound to `scrollOffset`.
- Same in `DefiMainScreen.swift:191-193`.

**Fix:** derive a 0…1 collapse progress from `scrollOffset` (e.g. last ~40pt before `contentInset`) and drive both balances from it with non-overlapping ranges — content fades out over the first half, header fades in over the second.

**Done when:** never two balances legible at once; transition scrubs with the drag and reverses cleanly; Wallet + DeFi, iOS + macOS, balance-hidden included.

### PR DESCRIPTION
**Pull request by gastonm5**
## Cause

Three things, all confirmed against `main`:

- **The threshold is binary and fires almost immediately.** `showBalanceInHeader = offset < contentInset` — and `contentInset` (78) is the scroll view's *own top inset*, which is what the geometry reader reports **at rest**. So the condition became true after ~1pt of scroll: the top bar started bringing its balance in while the large one had barely moved.
- **The top bar animated its own swap on a wall-clock spring** (`withAnimation(.interpolatingSpring)` inside `.onChange(of: showBalance)`). A 0.3s spring cannot stay in sync with a finger-driven translation, by construction.
- **The large balance had no scroll-driven opacity at all.** Nothing ever told `VaultMainBalanceView` to fade.

Result: mid-transition both balances are fully drawn — it reads as a duplicate, not a transition. `VaultMainScreen` and `DefiMainScreen` had the threshold **byte-identical**, so the Wallet and DeFi tabs had the same bug twice.

## The fix

A `HeaderCollapseProgress` value derived from the scroll offset, split into two **non-overlapping** fade ramps:

```
d       = restingOffset − offset            // points scrolled, 0 at rest
value   = clamp(d / 110, 0, 1)
expandedOpacity  = max(1 − value / 0.5, 0)          // 1 → 0 over value 0…0.5
collapsedOpacity = max((value − 0.5) / 0.5, 0)      // 0 → 1 over value 0.5…1
isCollapsed      = value >= 0.5
```

- The **large balance** and the **top bar's toolbar buttons** fade out over the first half.
- The **top bar's balance** and its opaque background fade in over the second half.
- The structural buttons ↔ balance swap happens at `value == 0.5` — the single point where *both* ramps are zero — so nothing pops in or out at the switch.

`expandedOpacity` and `collapsedOpacity` are never both non-zero. That is not a tuning outcome, it's an algebraic property of the split, and it's pinned by a test that sweeps the whole transition (plus 40pt past each end) at 0.25pt resolution. Overscroll, reverse scrolling and re-entry all fall out of it for free, because the whole thing is a pure function of the offset with no timeline anywhere.

## The shared extraction

Rather than fixing the threshold twice:

| New | |
|---|---|
| `HeaderCollapseProgress` | the pure value type above, `Equatable`, no SwiftUI |
| `HomeHeaderCollapse` | per-tab store, written by the scroll views, read by leaves |
| `headerCollapseFade` / `headerCollapseReveal` | the two `ViewModifier`s that apply the ramps |

Wallet and DeFi now share one implementation. `VaultMainScreen`, `DefiMainScreen`, `HomeScreen` and `HomeMainHeaderView` all drop the `showBalanceInHeader` boolean path, and `HomeScreen` loses `walletShowPortfolioHeader` / `defiShowPortfolioHeader` / `showPortfolioHeader` / `updateHeader()` entirely.

## Scroll performance — this screen was profiled, and this change is a net reduction

This is cosmetic polish landing on the screen [#4974](https://github.com/vultisig/vultisig-ios/issues/4974) took from **114 → 4.4 ms/s** hitch time (PRs #4975 / #4976), so "does it cost anything" was the design constraint, not an afterthought.

**It removes more work than it adds:**

- **Removed:** the scroll offset no longer travels through a `@Binding`. That binding wrote `@State` on `VaultMainScreen` / `DefiMainScreen` at 5 Hz for the *entire duration of every scroll*, re-running the screen body — the documented whole-content-tree invalidation. `ScrollOffsetThrottle` existed only to rate-limit that binding; with the binding gone it has nothing left to protect, so it goes with it. (Explicitly flagging this so it doesn't read as a silent revert of #4975's step 1 — it is the same defect being removed one level further up.)
- **Added, and it is bounded:** `HomeHeaderCollapse.update` compares before it publishes, and the progress is saturated at 0 or 1 everywhere outside a 110pt window. **The overwhelming majority of any scroll publishes nothing at all** — the per-layout-pass callback costs one subtraction, one divide, one clamp and one comparison, with zero view invalidations.
- **Inside the window,** the observers are leaves only: `HeaderCollapseFade` / `HeaderCollapseReveal` (each re-applies an `.opacity` to an already-built subtree — the balance view's own body does not re-run) and the bar's trailing slot. `HomeScreen` holds the store in a plain `@State` and never reads it; `VaultMainScreen` and `DefiMainScreen` hold it as a plain `let`. **No screen body, and no chain-list row, is re-evaluated by a collapse.**
- The top bar itself was deliberately kept *out* of the observation: `HomeMainHeaderView.balanceText` walks the vault's DeFi coins on the DeFi tab, so the trailing slot was split into its own view that receives the balance as a plain `String`. Otherwise every frame of a DeFi collapse would have redone that walk. (Caught by the cross-model review, not by me.)

**Honest limits:** this reasoning is static. The change was **not** re-profiled on-device with Instruments — that needs a real device and a large vault, per the protocol in the #4974 write-up. What I can say is that it deletes a continuous 5 Hz whole-tree invalidation and replaces it with something that is silent outside a 110pt window and touches only leaves inside it. If a reviewer disagrees with that analysis, the fix is not worth shipping unmeasured and I'd rather it be held.

## Verification

**Visual — recorded renders of the real views.** I composited the real `HomeMainHeaderView` over the real balance content exactly as `HomeScreen` stacks them, translated the content by a scroll distance and derived the progress from that same distance, then rendered a filmstrip at 0 / 25 / 50 / 75 / 100 % of the transition for **Wallet**, **Wallet with the balance hidden**, and **DeFi**. All 15 frames inspected:

| progress | what's on screen |
|---|---|
| 0 % | large balance at full opacity, toolbar buttons at full opacity, no bar background |
| 25 % | large balance ~50 %, buttons ~50 %, bar still shows no balance |
| 50 % | large balance gone, buttons gone, bar balance not yet in — the swap point, nothing drawn twice |
| 75 % | bar balance ~50 % with its background half in, large balance gone |
| 100 % | bar balance and background fully in, large balance gone |

The balance-hidden run behaves identically — the dots crossfade on the same ramps. The DeFi run fades the whole `DefiMainBalanceView` card the same way.

The harness that recorded these is **not committed** — reference PNGs in CI are a maintenance cost I didn't think this earned, and the non-overlap property is already covered deterministically by the unit tests.

**What I could not verify.** I have no vault with balances to drive the app, so I did not scrub this with a finger on a device or a simulator. Specifically unverified and worth a human minute:

1. **The feel of the 110pt distance** — it's a tuned constant, judged from static renders, not from dragging.
2. **Reversal and fling behaviour.** The maths reverses exactly (it's a pure function of the offset), but the *feel* of a fast fling through the window is not something a still frame shows.
3. **macOS.** The same views build for macOS and there is no platform branch in any of this beyond the header's existing `isMacOS` padding and `RefreshToolbarButton`, but it was rendered on iOS only.
4. **Switching between a scrolled tab and an unscrolled one** still overlaps briefly, because `TabView` cross-fades the two pages and the outgoing page's balance is still on screen. That is the tab-switch transition rather than the scroll-driven collapse, it predates this change, and I left it alone deliberately — the top bar's own animation was removed so it no longer contributes.

**Gate.** `xcodebuild test` (full suite, iOS Simulator, dedicated en_US sim): `** TEST SUCCEEDED **`, **4626 tests, 1 skipped, 0 failures**. SwiftLint: 39 warnings / 0 errors repo-wide, **none in any file this touches** (all pre-existing). `make generate` clean.

**Cross-model review.** Every commit reviewed by `codex exec --sandbox read-only` before the next one started, plus a whole-branch pass. Two MAJORs found and fixed (the tab-switch animation re-creating the overlap; the DeFi balance walk running per transition frame), one MINOR rejected with reason.

## Coverage

- **Wallet tab** ✅ — `VaultMainBalanceView` fades, bar balance crossfades.
- **DeFi tab** ✅ — same modifier, same progress, `DefiMainBalanceView` fades as a card.
- **iOS** ✅ rendered · **macOS** ⚠️ builds and shares the code path, not rendered.
- **Balance hidden** ✅ — `String.hideBalanceText` dots ride the identical ramps.

Closes #5013

## Summary by CodeRabbit

* **New Features**
  * Added coordinated header collapse behavior across wallet and DeFi screens.
  * Header content now smoothly fades between expanded and collapsed states as users scroll.
  * Wallet and DeFi tabs retain independent header positions.
  * Collapsed header controls and balance information transition more consistently.

* **Bug Fixes**
  * Improved scroll-based header updates and reduced unnecessary visual updates.
  * Camera tab behavior remains unaffected by header collapse changes.

* **Tests**
  * Added coverage for collapse progress, opacity transitions, tab independence, and edge cases.

### CODE REVIEW

File: VultisigApp/VultisigApp/Features/Wallet/VaultMain/Views/HomeMainHeaderView.swift [hunk 1]
```diff
@@ -56,27 +60,66 @@ struct HomeMainHeaderView: View {
...
+    /// so the swap lands on the one point where both are invisible.
+    var body: some View {
+        if progress.isCollapsed {
             balanceView
+                .opacity(progress.collapsedOpacity)
         } else {
             buttonsStack
+                .opacity(progress.expandedOpacity)
         }
     }
```
[the remark this answers is not in the document]
Comment (gastonm5) [reply VultisigApp/VultisigApp/Features/Wallet/VaultMain/Views/HomeMainHeaderView.swift:126]: Accepted, and fixed a level up in `8d4507bc1`.

You're right that the swap was unguarded, and it was an internal inconsistency rather than an oversight in isolation: `HeaderCollapseFade` already carried exactly this guard for the large balance, so the pattern existed and this one call site didn't get it.

Rather than open-code the two lines a third time, I lifted the guard into a shared `headerCollapseOpacity(_:)` helper in `HeaderCollapseModifiers.swift` and routed every ramped view through it — the large balance, the top bar's background, and both sides of the trailing swap:

```swift
func headerCollapseOpacity(_ opacity: Double) -> some View {
    self
        .opacity(opacity)
        .allowsHitTesting(opacity > 0)
        .accessibilityHidden(opacity == 0)
}
```

Applying a ramp and forgetting the guard is no longer something a call site can do, which removes the class rather than this instance. The bar's background picked the guard up too: it is removed outright while the home is expanded, so it was already covered everywhere except the single position where it is present at zero opacity.

One thing worth noting for anyone reading later: both guards compare against zero *exactly*, so the tests now pin that the ramps land on hard zeros at both ends (they do — the `max(..., 0)` clamps guarantee it) and that `opacity > 0` and `opacity == 0` partition the whole sweep. A ramp that ever went negative would satisfy neither and leave a view invisible *and* still exposed, which is precisely the defect the guards exist to prevent.

Verified: full unit suite green on iPhone 16 Pro — 4629 tests, 1 skipped, 0 failures, `** TEST SUCCEEDED **`. SwiftLint unchanged at the repo baseline of 39 violations. At scroll rest the progress is 0, so the History/Settings identifiers the XCUITests drive stay reachable.

**Review by johnnyluo — APPROVED** [review]
LGTM
